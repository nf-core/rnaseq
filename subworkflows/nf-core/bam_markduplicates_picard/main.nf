//
// Picard MarkDuplicates, index BAM file and run samtools stats, flagstat and idxstats
//

include { PICARD_MARKDUPLICATES } from '../../../modules/nf-core/picard/markduplicates/main'
include { SAMTOOLS_INDEX        } from '../../../modules/nf-core/samtools/index/main'
include { BAM_STATS_SAMTOOLS    } from '../bam_stats_samtools/main'

workflow BAM_MARKDUPLICATES_PICARD {

    take:
    ch_bam   // channel: [ val(meta), path(bam) ]
    ch_fasta // channel: [ path(fasta) ]
    ch_fai   // channel: [ path(fai) ]

    main:

    ch_versions = Channel.empty()

    PICARD_MARKDUPLICATES ( ch_bam, ch_fasta, ch_fai )
    ch_versions = ch_versions.mix(PICARD_MARKDUPLICATES.out.versions.first())

    SAMTOOLS_INDEX ( PICARD_MARKDUPLICATES.out.bam )
    ch_versions = ch_versions.mix(SAMTOOLS_INDEX.out.versions.first())

    ch_bam_bai = PICARD_MARKDUPLICATES.out.bam
        .join(SAMTOOLS_INDEX.out.bai, by: [0], remainder: true)
        .join(SAMTOOLS_INDEX.out.csi, by: [0], remainder: true)
        .map{meta, bam, bai, csi ->
            if (bai) [ meta, bam, bai ]
            else [ meta, bam, csi ]
        }

    BAM_STATS_SAMTOOLS ( ch_bam_bai, ch_fasta )
    ch_versions = ch_versions.mix(BAM_STATS_SAMTOOLS.out.versions)

    emit:
    bam      = PICARD_MARKDUPLICATES.out.bam     // channel: [ val(meta), path(bam) ]
    metrics  = PICARD_MARKDUPLICATES.out.metrics // channel: [ val(meta), path(bam) ]
    bai      = SAMTOOLS_INDEX.out.bai            // channel: [ val(meta), path(bai) ]
    csi      = SAMTOOLS_INDEX.out.csi            // channel: [ val(meta), path(csi) ]

    stats    = BAM_STATS_SAMTOOLS.out.stats      // channel: [ val(meta), path(stats) ]
    flagstat = BAM_STATS_SAMTOOLS.out.flagstat   // channel: [ val(meta), path(flagstat) ]
    idxstats = BAM_STATS_SAMTOOLS.out.idxstats   // channel: [ val(meta), path(idxstats) ]

    versions = ch_versions                       // channel: [ versions.yml ]
}
