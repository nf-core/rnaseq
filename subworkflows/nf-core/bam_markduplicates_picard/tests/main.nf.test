nextflow_workflow {

    name "Test Workflow BAM_MARKDUPLICATES_PICARD"
    script "../main.nf"
    workflow "BAM_MARKDUPLICATES_PICARD"

    tag "picard"
    tag "picard/markduplicates"
    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "bam_markduplicates_picard"
    tag "subworkflows/bam_markduplicates_picard"
    tag "subworkflows/bam_stats_samtools"
    tag "bam_stats_samtools"
    tag "samtools"
    tag "samtools/flagstat"
    tag "samtools/idxstats"
    tag "samtools/index"
    tag "samtools/stats"

    test("sarscov2 - bam") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end: false ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true)
                ])
                input[1] = Channel.of([
                    [ id:'genome' ],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ])
                input[2] = Channel.of([
                    [ id:'genome' ],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta.fai', checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    path(workflow.out.bam[0][1]),
                    path(workflow.out.bai[0][1]),
                    path(workflow.out.flagstat[0][1]),
                    path(workflow.out.idxstats[0][1]),
                    path(workflow.out.stats[0][1]),
                ).match("sarscov2 - bam")  },
                { assert path(workflow.out.metrics.get(0).get(1)).getText().contains("97") }
            )
        }
    }

    test("homo_sapiens - cram") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/cram/test.paired_end.sorted.cram', checkIfExists: true)
                ])
                input[1] = Channel.of([
                    [ id:'genome' ],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ])
                input[2] = Channel.of([
                    [ id:'genome' ],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    path(workflow.out.bam[0][1]),
                    path(workflow.out.bai[0][1]),
                    path(workflow.out.flagstat[0][1]),
                    path(workflow.out.idxstats[0][1]),
                    path(workflow.out.stats[0][1]),
                ).match("homo_sapiens - cram")  },
                { assert path(workflow.out.metrics.get(0).get(1)).getText().contains("0.999986") }
            )
        }
    }

}
