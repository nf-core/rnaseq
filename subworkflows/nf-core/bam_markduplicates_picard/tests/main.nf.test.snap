{
    "homo_sapiens - cram": {
        "content": [
            "test.bam:md5,6641dc05efa8384a061f378d86d922cd",
            "test.bam.bai:md5,c41c60d8a94adebe53b6df80b6e90d38",
            "test.flagstat:md5,93b0ef463df947ede1f42ff60396c34d",
            "test.idxstats:md5,e179601fa7b8ebce81ac3765206f6c15",
            "test.stats:md5,9ac28e327a7797d7bb6a5922fde59ed1"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.01.0"
        },
        "timestamp": "2024-02-13T16:46:16.302755774"
    },
    "sarscov2 - bam": {
        "content": [
            "test.bam:md5,3091fe6ba1b7530f382fe40b9fd8f45b",
            "test.bam.bai:md5,4d3ae8d013444b55e17aa0149a2ab404",
            "test.flagstat:md5,4f7ffd1e6a5e85524d443209ac97d783",
            "test.idxstats:md5,df60a8c8d6621100d05178c93fb053a2",
            "test.stats:md5,d7796222a087b9bb97f631f1c21b9c95"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.01.0"
        },
        "timestamp": "2024-02-13T16:46:02.942115679"
    }
}